feat: swap websocket transport with cdp add binding/evaluate

ryanthemanuel opened the pull request:
### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

This PR:

* Swaps out socket io transport with CDP transport on chromium browsers.
* Cleans up some of the CDP connect/disconnect logic

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

This can be tested by just running Cypress to ensure that things worked as previously

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

n/a

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [X] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

ryanthemanuel committed:
let's do this -- run ci

ryanthemanuel committed:
try testing part 2

ryanthemanuel committed:
disable auto cancellation

ryanthemanuel committed:
empty commit -- run ci

ryanthemanuel committed:
let's see how much farther we get

ryanthemanuel committed:
empty commit -- run ci

ryanthemanuel committed:
empty commit -- run ci

ryanthemanuel committed:
empty commit -- run ci

ryanthemanuel committed:
empty commit -- run ci

ryanthemanuel committed:
split out ios -- run ci

ryanthemanuel committed:
cy in cy -- run ci

ryanthemanuel committed:
fix some tests -- run ci

ryanthemanuel committed:
fix some tests -- run ci

ryanthemanuel committed:
update interface -- run ci

ryanthemanuel committed:
update interface -- run ci

ryanthemanuel committed:
use socket io parser -- run ci

ryanthemanuel committed:
undo refactor -- run ci

ryanthemanuel committed:
fix clean up

ryanthemanuel committed:
fix some tests -- run ci

ryanthemanuel committed:
better cleanup -- run ci

ryanthemanuel committed:
fix build

ryanthemanuel committed:
fix some tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
fix tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
skip tests -- run ci

ryanthemanuel committed:
try to recreate with debugging

ryanthemanuel committed:
try and await -- run ci

ryanthemanuel committed:
try and await -- run ci

ryanthemanuel committed:
try and await -- run ci

ryanthemanuel committed:
Update cri-client.ts

ryanthemanuel committed:
try and await -- run ci

ryanthemanuel committed:
try and await -- run ci

ryanthemanuel committed:
better impl

ryanthemanuel committed:
better impl

ryanthemanuel committed:
back out experiment

ryanthemanuel committed:
binary data

ryanthemanuel committed:
unskip tests

ryanthemanuel committed:
fix crash on reconnect

ryanthemanuel committed:
fix integration test

ryanthemanuel committed:
add debugging

brian-mann committed:
fix user-agent not being available when using the CDPSocket

- add coreData.app.browserUserAgent getters + setters
- move logic to be set through the clientRoute http request

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

brian-mann committed:
move setBrowserStatus to a data-context action since it's a mutation

brian-mann committed:
sort import list

brian-mann committed:
remove lots of indirection that was hoisting arbitrary getters to data context

brian-mann committed:
properly update projects with ctx.update, remove unused private getter

brian-mann committed:
refactor server actions into their own data context class

brian-mann committed:
fix integration tests with the userAgent change

brian-mann committed:
fix property accessors

brian-mann committed:
update setters to use actions

brian-mann committed:
cleanup function refs

brian-mann committed:
invoke the promisified destroy function

brian-mann committed:
update debug statements to log out the target

ryanthemanuel committed:
improve cleanup

ryanthemanuel committed:
fix test

ryanthemanuel committed:
fix tests

ryanthemanuel committed:
fix next tick issue

ryanthemanuel committed:
add debugging

ryanthemanuel committed:
try and fix clean up logic

ryanthemanuel committed:
fix electron

ryanthemanuel committed:
Merge branch 'release/13.0.0' into ryanm/feat/swap-websockets-with-cdp

ryanthemanuel committed:
add debuggin

ryanthemanuel committed:
debugging

brian-mann committed:
use single line options debug

brian-mann committed:
add missing system test for when the tab/page is closed

brian-mann committed:
ensure verbose debug logs when cdp msgs are received are logged first before reactions to the events

brian-mann committed:
prevent race conditional calling close twice hangs the close promise

- add more debug logs for closing cri clients
- make cri.close is async even when we early exit

brian-mann committed:
add tests for handling unexpected Browser.close events

brian-mann committed:
refactor connecting to CDP socket, encapsulate local variables

brian-mann committed:
add new BrowserCriClient state to manage unexpected crashes for the browser or page

brian-mann committed:
temporarily comment out reconnecting on a disconnect

brian-mann committed:
begin implementation handling recovery from unexpected browser or page closes

- TODO on browser close, do not continue running tests, abort the run
- TODO on page close, continue running tests

brian-mann committed:
exit the run completely when the browser process closes

brian-mann committed:
add error snapshots

ryanthemanuel committed:
fix origin tests

ryanthemanuel committed:
fix bluebird

ryanthemanuel committed:
add test for page closed - chrome

ryanthemanuel committed:
fix firefox

ryanthemanuel committed:
slight refactor

brian-mann committed:
rework logic to create signals for whether a graceful shutdown is happening

- intention is to make the chrome/electron tab/browser process closed logic work cohesively together

ryanthemanuel committed:
retries

brian-mann committed:
do not check for browser/page closes when targetDestroyed is for a previously disposed page

ryanthemanuel commented on .circleci/workflows.yml:
```diff
@@ -508,7 +508,7 @@ commands:
               CYPRESS_RECORD_KEY=$MAIN_RECORD_KEY \
               CYPRESS_INTERNAL_ENABLE_TELEMETRY="true" \
-              yarn cypress:run --record --parallel --group 5x-driver-<<parameters.browser>> --browser <<parameters.browser>> --runner-ui
+              DEBUG=cypress:server:browsers*,cypress:server:run yarn cypress:run --auto-cancel-after-failures false --record --parallel --group 5x-driver-<<parameters.browser>> --browser <<parameters.browser>> --runner-ui
```
Reminder to remove these.

ryanthemanuel committed:
code cleanup

ryanthemanuel committed:
minor rework

ryanthemanuel committed:
fix unit tests

ryanthemanuel committed:
Update browser-cri-client.ts

ryanthemanuel committed:
fix issue with hanging

ryanthemanuel commented on packages/socket/lib/utils.ts:
```diff
@@ -0,0 +1,39 @@
+// @ts-ignore
+import * as engineParser from 'engine.io-parser'
+
+export const encode = (data: any, namespace: string) => {
```
Type the return type of this.

ryanthemanuel replied:
Updated: [`4e1085d` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/4e1085dc7cbee1be2225ce93bccbcac79deafe8a)

ryanthemanuel commented on packages/socket/lib/utils.ts:
```diff
@@ -0,0 +1,39 @@
+  })
+}
+
+export const decode = (data: any) => {
```
Type the return type

ryanthemanuel replied:
Updated: [`4e1085d` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/4e1085dc7cbee1be2225ce93bccbcac79deafe8a)

ryanthemanuel committed:
minor cleanup

mjhenkes commented on packages/errors/__snapshot-html__/VIDEO_UPLOAD_ON_PASSES_REMOVED.html:
```diff
@@ -0,0 +1,42 @@
+<!DOCTYPE html>
```
Is this snapshot new? i don't see the `VIDEO_UPLOAD_ON_PASSES_REMOVED` string added below.

ryanthemanuel replied:
It's not new from this PR. This was a previously added error that got missed along the way somehow.

mjhenkes commented on packages/server/lib/browsers/cri-client.ts:
```diff
@@ -143,20 +171,21 @@ export const create = async (target: string, onAsynchronousError: Function, host
-    // @see https://github.com/cyrus-and/chrome-remote-interface/issues/72
-    cri._notifier.on('disconnect', reconnect)
+    // TODO: add this back in
+    // cri.on('disconnect', reconnect)
```
TODO, just calling out this todo so it doesn't get dropped.

ryanthemanuel replied:
Added back: [`4e1085d` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/4e1085dc7cbee1be2225ce93bccbcac79deafe8a)

mjhenkes commented on packages/server/lib/browsers/electron.ts:
```diff
@@ -287,18 +287,19 @@ export = {
     await cdpAutomation._handlePausedRequests(browserCriClient?.currentlyAttachedTarget)
     cdpAutomation._listenForFrameTreeChanges(browserCriClient?.currentlyAttachedTarget)
 
+    await win.loadURL(url)
```
I think you mentioned this was a bug fix, could you throw a comment in here about why this should be called after?

ryanthemanuel replied:
Added: [`faf7cfe` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/faf7cfe3e19bb11a999213dd8efeb2056ede2409)

mjhenkes commented on packages/server/lib/request.js:
```diff
@@ -664,8 +662,8 @@ module.exports = function (options = {}) {
-    sendPromise (headers, automationFn, options = {}) {
-      let a; let c; let ua
+    sendPromise (userAgent, automationFn, options = {}) {
+      let a; let c
```
could we use more descriptive variable names here?

ryanthemanuel replied:
Updated: [`4e1085d` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/4e1085dc7cbee1be2225ce93bccbcac79deafe8a)

mjhenkes commented on packages/socket/lib/cdp-browser.ts:
```diff
@@ -0,0 +1,65 @@
+
+    // Set timeout so that the connect event is emitted after the constructor returns and the user has a chance to attach a listener
+    setTimeout(() => {
+      super.emit('connect')
```
do we want to make a connect call to the server to ensure the connection is good before assuming we've connected?

ryanthemanuel replied:
We shouldn't need to.  This code gets added with Cypress.  All of the add bindings/scripts/etc. happen prior to any pages loading, so everything should be ready

mjhenkes commented on packages/socket/lib/cdp-browser.ts:
```diff
@@ -0,0 +1,65 @@
+  emit = (event: string, ...args: any[]) => {
+    // Generate a unique key for this event
+    const uuid = uuidv4()
+    const key = `${event}-${uuid}`
```
could the key just be the uuid?

ryanthemanuel replied:
Gonna give this a try.

ryanthemanuel replied:
Updated: [`4e1085d` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/4e1085dc7cbee1be2225ce93bccbcac79deafe8a)

mjhenkes commented on packages/socket/lib/cdp-socket.ts:
```diff
@@ -0,0 +1,180 @@
+    encode([event, callbackEvent, args], this._namespace).then((encoded: any) => {
+      const expression = `
+        if (window['cypressSocket-${this._namespace}'] && window['cypressSocket-${this._namespace}'].send) {
+          window['cypressSocket-${this._namespace}'].send('${JSON.stringify(encoded).replaceAll('\\', '\\\\').replaceAll('\'', '\\\'')}')
```
Wondering aloud if there is a better way to do this... I did this originally and it seems to work... but it sure is ugly.

ryanthemanuel replied:
I tried a few things and couldn't find a way. Will keep thinking about it though.

ryanthemanuel committed:
PR comments

ryanthemanuel committed:
Update electron.ts

ryanthemanuel committed:
Update electron.ts

ryanthemanuel committed:
fix retry issue

ryanthemanuel committed:
fix retry issue

ryanthemanuel committed:
revert

ryanthemanuel committed:
fix reconnect issue

ryanthemanuel committed:
fix retry

brian-mann committed:
remove test for electron browser crashing bc its invalid

ryanthemanuel committed:
Update electron.ts

ryanthemanuel committed:
fix cy in cy problems

ryanthemanuel committed:
fix cy in cy problems

ryanthemanuel commented on packages/server/lib/browsers/electron.ts:
```diff
@@ -53,16 +53,22 @@ const _getAutomation = async function (win, options: BrowserLaunchOpts, parent)
 
   const pageCriClient = await browserCriClient.attachToTargetUrl('about:blank')
 
   const sendClose = () => {
```
```suggestion
  const sendClose = async () => {
```

ryanthemanuel committed:
Update packages/server/lib/browsers/electron.ts

ryanthemanuel committed:
Merge remote-tracking branch 'origin/ryanm/feat/swap-websockets-with-cdp' into ryanm/test/sockets-cdp

ryanthemanuel committed:
fix unit tests

ryanthemanuel commented on packages/server/lib/browsers/electron.ts:
```diff
@@ -283,21 +295,26 @@ export = {
     await this._enableDebugger()
 
-    await win.loadURL(url)
+    if (cdpAutomation) {
```
```suggestion
    // Note that these calls have to happen before we load the page so that we don't miss out on any events that happen quickly
    if (cdpAutomation) {
```

ryanthemanuel committed:
update workflows.yml

ryanthemanuel commented on .circleci/workflows.yml:
```diff
@@ -653,7 +653,7 @@ commands:
             SPECS=`echo $SPECS | xargs -n 1 | circleci tests split --split-by=timings`
             echo SPECS=$SPECS
-            yarn workspace @tooling/system-tests test:ci $SPECS --browser <<parameters.browser>>
+            DEBUG=cypress*:server:browsers:cri-client:* yarn workspace @tooling/system-tests test:ci $SPECS --browser <<parameters.browser>>
```
```suggestion
            yarn workspace @tooling/system-tests test:ci $SPECS --browser <<parameters.browser>>
```

ryanthemanuel committed:
Apply suggestions from code review

ryanthemanuel committed:
Merge branch 'release/13.0.0' into ryanm/feat/swap-websockets-with-cdp

ryanthemanuel committed:
experiment

ryanthemanuel committed:
merge release/13.0.0

brian-mann committed:
add integration tests for cdp socket reconnection

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

ryanthemanuel commented on packages/socket/lib/cdp-socket.ts:
```diff
@@ -0,0 +1,175 @@
+    })
+  }
+
+  // TODO: figure out end lifecycle/disconnects/etc.
```
```suggestion
```

ryanthemanuel committed:
Update packages/socket/lib/cdp-socket.ts

ryanthemanuel commented on packages/socket/lib/utils.ts:
```diff
@@ -0,0 +1,39 @@
+// @ts-ignore
+import * as engineParser from 'engine.io-parser'
+
+export const encode = (data: any, namespace: string): Promise<string> => {
```
Add some error handling

chrisbreiding replied:
Are you planning to add error handling each of the places that `encode` and `decode` are used?

ryanthemanuel replied:
Yeah I was trying to think through that. I'm not sure what else we can do other than log. I'll push that for now, but I may chat with @brian-mann to see if he has additional thoughts

ryanthemanuel replied:
[`b4c9b4c` (#27592)](https://github.com/cypress-io/cypress/pull/27592/commits/b4c9b4c22c0a91a1d5da5e68b86c7ab39f2843c2)

ryanthemanuel replied:
Spoke with @brian-mann offline. We're going to let these bubble as they would be errors that are unrecoverable.

brian-mann committed:
add tests for stopping the reconnection after close is called

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

ryanthemanuel committed:
add error handling

brian-mann committed:
when the spec closes in run mode always abort the run

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

brian-mann committed:
support overriding system test timeout with env var

ryanthemanuel committed:
update with debug logs on errors

brian-mann committed:
update error message on browser/tab crash to clarify which browserName crashed, indicate we're continuing running the next spec

brian-mann committed:
remove unnecessary handler on Target.targetCrashed

this was preventing crashed tabs from being closed when moving to the next spec

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

brian-mann committed:
remove test for electron crashing bc it hangs the entire cypress process and appears to be invalid

ryanthemanuel committed:
let unhandled unrecoverable errors bubble up

brian-mann committed:
add note about why we aren't testing browser crashes in electron [skip ci]

brian-mann committed:
Merge branch 'ryanm/feat/swap-websockets-with-cdp' of https://github.com/cypress-io/cypress into ryanm/feat/swap-websockets-with-cdp

brian-mann committed:
fix errors not exiting on uncaught exception, fixed error

ryanthemanuel committed:
fix typescript

ryanthemanuel committed:
Merge branch 'release/13.0.0' into ryanm/feat/swap-websockets-with-cdp

ryanthemanuel committed:
fix builds

ryanthemanuel committed:
fix test
